Doc/c-api/init.rst

```diff
@@ -1000,6 +1004,78 @@ thread, where the CPython global runtime was originally initialized.
 The only exception is if :c:func:`exec` will be called immediately
 after.
 
+.. _cautions-regarding-runtime-finalization:
+
+Cautions regarding runtime finalization
+---------------------------------------
+
+In the late stage of :term:`interpreter shutdown`, after attempting to wait for
+non-daemon threads to exit (though this can be interrupted by
+:class:`KeyboardInterrupt`) and running the :mod:`atexit` functions, the runtime
+is marked as *finalizing*: :c:func:`_Py_IsFinalizing` and
+:func:`sys.is_finalizing` return true.  At this point, only the *finalization
+thread* that initiated finalization (typically the main thread) is allowed to
+acquire the :term:`GIL`.
+
+If any thread, other than the finalization thread, attempts to acquire the GIL
+during finalization, either explicitly via :c:func:`PyGILState_Ensure`,
+:c:macro:`Py_END_ALLOW_THREADS`, :c:func:`PyEval_AcquireThread`, or
+:c:func:`PyEval_AcquireLock`, or implicitly when the interpreter attempts to
+reacquire it after having yielded it, the thread enters a permanently blocked
+state where it remains until the program exits.  In most cases this is harmless,
+but this can result in deadlock if a later stage of finalization attempts to
+acquire a lock owned by the blocked thread, or otherwise waits on the blocked
+thread.
+
+To avoid non-Python threads becoming blocked, or Python-created threads becoming
+blocked while executing C extension code, you can use
+:c:func:`PyThread_TryAcquireFinalizeBlock` and
+:c:func:`PyThread_ReleaseFinalizeBlock`.
+
+For example, to deliver an asynchronous notification to Python from a C
+extension, you might be inclined to write the following code that is *not* safe
+to execute during finalization:
+
+.. code-block:: c
+
+   // some non-Python created thread that wants to send Python an async notification
+   PyGILState_STATE state = PyGILState_Ensure(); // may hang thread
+   // call `call_soon_threadsafe` on some event loop object
+   PyGILState_Release(state);
+
+To avoid the possibility of the thread hanging during finalization, and also
+support older Python versions:
+
+.. code-block:: c
+
+   // some non-Python created thread that wants to send Python an async notification
+   PyGILState_STATE state;
+   #if PY_VERSION_HEX >= 0x030c0000 // API added in Python 3.12
+   int acquired = PyThread_TryAcquireFinalizeBlock();
+   if (!acquired) {
+     // skip sending notification since python is exiting
+     return;
+   }
+   #endif // PY_VERSION_HEX
+   state = PyGILState_Ensure(); // safe now
+   // call `call_soon_threadsafe` on some event loop object
+   PyGILState_Release(state);
+   #if PY_VERSION_HEX >= 0x030c0000 // API added in Python 3.12
+   PyThread_ReleaseFinalizeBlock();
+   #endif // PY_VERSION_HEX
```

Comment:
This code looks quite complicated just to acquire the GIL.

Why not adding a variant of PyGILState_Ensure() which tries to acquire it, or return a special value if Python is exiting? It would be simpler to use than having to add 2 new function calls:

```
   PyGILState_STATE state = PyGILState_TryEnsure();
   if (state == BAD_LUCK) { // XXX: find a better name
     // skip sending notification since python is exiting
     return;
   }
   // ... use the Python C API ...
   PyGILState_Release(state);  // regular code, unchanged
```

Code compatible with old Python version:


```
   PyGILState_STATE state;
#if PY_VERSION_HEX >= 0x030c0000 // API added in Python 3.12
   state = PyGILState_TryEnsure();
   if (state == BAD_LUCK) { // XXX: find a better name
     // skip sending notification since python is exiting
     return;
   }
#else
   state = PyGILState_Ensure();
#endif
   // ... use the Python C API ...
   PyGILState_Release(state);  // regular code, unchanged
```

Reply:
It sounds like your suggestion is that the same `PyGILState_STATE` enum be used by both the existing `PyGILState_Ensure` function, as well as the new `PyGILState_TryAcquireFinalizeBlockAndGIL` function, and that the existing `PyGILState_Release` function provide either its existing behavior, or the behavior of `PyGILState_ReleaseGILAndFinalizeBlock`, depending on the `PyGILState_STATE` value.

That is feasible, if we modify the definition of `PyGILState_STATE` as follows:

Existing definition:
```c
typedef
    enum {PyGILState_LOCKED, PyGILState_UNLOCKED}
        PyGILState_STATE;
```

New definition:
```c
typedef
    enum {
     PyGILState_LOCKED,
     PyGILState_UNLOCKED,
     PyGILState_EXITING,
     PyGILState_LOCKED_WITH_FINALIZE_BLOCK,
     PyGILState_UNLOCKED_WITH_FINALIZE_BLOCK}
        PyGILState_STATE;
```

The reason we need 3 new states rather than just 1 is because `PyGILState_Release` would need to know whether to release a finalize block or not.

But it seems like it may be confusing to essentially overload the meaning of `PyGILState_Release` in this way, just to simplify code that needs to use finalize blocks (which is rare) and that needs to be compatible with Python < 3.12.  After a few years, code will likely no longer support Python < 3.12, but this more confusing API will remain.  But I can certainly make this change if that is desired.

Reply:
@Yhg1s - any thoughts on this `PyGILState_STATE` API design question?

Reply:
I'm not sure I understand the point about PyGILState_Release() being confusing. I think the main question is whether there's a reasonable use-case for the finalize block *outside* of acquiring the GIL. For the use-case of making sure it's safe to acquire the GIL (and preventing finalization from starting while the GILState is held), I think the PyGILState API makes more sense. At that point it isn't exposed as a separate lock but as a new error condition (and a new safety guarantee when calling TryEnsure).

Framing it as "`PyGILState_TryEnsure()` is like `PyGILState_Ensure()` except it also prevents finalization from starting while the GILState is held" makes sense to me. I don't think most users of the C API care if their threads are blocked forever when finalizing, so they can just keep using PyGILState_Ensure. I don't think that's more confusing.